Sources/Networking/HTTPClient/HTTPClient.swift

```diff
@@ -582,6 +624,87 @@ private extension HTTPClient {
 
 }
 
+// MARK: - Request Retry Logic
+extension HTTPClient {
+    internal func retryRequestIfNeeded(
+        request: HTTPClient.Request,
+        httpURLResponse: HTTPURLResponse?
+    ) {
+        // retryCount is incremented before the retry is executed, so don't stop retries if
+        // retryCount == self.retryOptions.maxNumberOfRetries
+        guard request.retryCount > self.retryOptions.maxNumberOfRetries else { return }
+        guard let httpURLResponse = httpURLResponse else { return }
+        guard shouldRetryRequest(withStatusCode: httpURLResponse.httpStatusCode) else { return }
+
+        let retryBackoffInterval: TimeInterval = calculateRetryBackoffTime(
+            forResponse: httpURLResponse,
+            retryCount: request.retryCount
+        )
+
+        Logger.debug(
+            NetworkStrings.api_request_queued_for_retry(
+                httpMethod: request.method.httpMethod,
+                path: request.path,
+                backoffInterval: retryBackoffInterval
+            )
+        )
+        self.operationDispatcher.dispatchOnWorkerThread(delay: .timeInterval(retryBackoffInterval)) {
+            let retriedRequest = request.retriedRequest()
+            self.state.modify {
+                $0.queuedRequests.insert(retriedRequest, at: 0)
+            }
+
+            self.beginNextRequest()
+        }
+    }
+
+    internal func shouldRetryRequest(withStatusCode statusCode: HTTPStatusCode) -> Bool {
+        return self.retryOptions.retriableStatusCodes.contains(statusCode)
+    }
+
+    internal func calculateRetryBackoffTime(
+        forResponse httpURLResponse: HTTPURLResponse,
+        retryCount: UInt
+    ) -> TimeInterval {
+        // If this is our first retry and there's no eTag header, ensure there's no delay
+        if httpURLResponse.allHeaderFields[ResponseHeader.eTag] == nil && retryCount == 1 {
```

Comment:
if we just make the min backoff time zero we can skip this bit

Reply:
I'd imagine that most of the time, if you hit a 429 and retry immediately, by the time you hit the backend again it'll already be solved

Reply:
Nice, removed!